ocaml/xapi/binpack.ml

```diff
@@ -171,6 +172,26 @@ let simulate_failure config dead_host =
   let placement = List.map (fun (vm, host) -> vm, (if List.mem_assoc vm plan then assoc "simulate_failure" vm plan else host)) config.placement in
   { config with hosts = hosts; placement = placement; num_failures = config.num_failures - 1 }
 
+let simulate_failure_approximation config =
+  let sum = List.fold_left (+) 0 in
+  let rec drop n = function
+    | [] -> []
+    | x::xl when n>0 -> drop (n-1) xl
+    | l -> l in
+
+  (* Assume all VMs are as big as the biggest *)
+  let max_vm_size = List.fold_left (fun acc (_, size) -> max acc size) 0L config.vms in
+
+  (* How many VMs can each host support to power on, ascending order *)
+  let hosts_by_capacity = List.sort compare (List.map (fun (_, h_size) -> Int64.(to_int (div h_size max_vm_size))) config.hosts) in
+  (* Assume biggest hosts fail, get left capacity *)
+  let left_capacity = drop config.num_failures hosts_by_capacity |> sum in
```

Comment:
Is this correct? When `hosts_by_capacity` is in ascending order, this drops the `num_failure` hosts with the smallest capacity. I was expecting:

```
let left_capacity = List.rev hosts_by_capacity |> drop config.num_failures |> sum in
```

This could be also expressed using `take` but because one would need `List.length hosts_by_capacity` this wouldn't be much more efficient.

Reply:
Oh, Thanks.
Now change it to descending order:
```
  (* How many VMs can each host support to power on, descending order *)
  let hosts_by_capacity = List.sort less_than' (List.map (fun (_, h_size) -> Int64.(to_int (div h_size max_vm_size))) config.hosts) in
  (* Assume biggest hosts fail, get left capacity *)
  let left_capacity = drop config.num_failures hosts_by_capacity |> sum in
```